### PR TITLE
Add a way to log hot threads in plain text

### LINKED ISSUE #110915: Improve the way we log hot threads
### Description

Certain scenarios (such as [LagDetector](https://www.elastic.co/guide/en/elasticsearch/reference/current/troubleshooting-unstable-cluster.html#_diagnosing_lagging_nodes_2)) are triggering elasticsearch to log its hot threads output. This helps us to understand what was elasticsearch doing at the moment and what was the bottleneck. However, the current output is not easy to use (the output is gzipped, base64 encodded and split into chunks). It requires some time to export and decode.

Instead we could use a plain text format that outputs each thread into a different log message (with a common correlation id).
This will output more log messages, however they would be easier to use and search as well.

HotThread collects 10 thread dump samples during 500ms. This might not be needed when trying to identify a single long running operation. We could instead collect and log only a single sample at the moment of request.

### PR DESCRIPTION
**Pull request by idegtiarenko**
This adds a way to log current threads in plain text to logs. This way we do not need to decode them and can search by stack trace in logs (for example to know if the issue is recurring). Please note, this produces a multi-line log entry.

Closes: #110915

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/monitor/jvm/HotThreads.java [hunk 1]
```diff
@@ -244,13 +277,15 @@ Map<Long, ThreadTimeAccumulator> getAllValidThreadInfos(ThreadMXBean threadBean,
...
-            // been started and has not yet died.
+
+        // NOTE, javadoc of getThreadInfo says: If a thread of the given ID is not alive or does not exist,
+        // null will be set in the corresponding element in the returned array. A thread is alive if it has
+        // been started and has not yet died.
+        for (int j = 0; j < threadElementsSnapshotCount - 1; j++) {
             result[j] = threadBean.getThreadInfo(threadIds, Integer.MAX_VALUE);
             Thread.sleep(threadElementsSnapshotDelay.millis());
         }
+        result[threadElementsSnapshotCount - 1] = threadBean.getThreadInfo(threadIds, Integer.MAX_VALUE);
```
Comment (idegtiarenko) [inline server/src/main/java/org/elasticsearch/monitor/jvm/HotThreads.java:288]: This avoids sleep after collecting the last sample.
Should make collection much faster when executing with `threadElementsSnapshotCount=1`

File: server/src/main/java/org/elasticsearch/monitor/jvm/HotThreads.java [hunk 2]
```diff
@@ -104,6 +105,34 @@ public static void logLocalHotThreads(Logger logger, Level level, String prefix,
         }
     }
 
+    /**
+     * Capture and log the current threads on the local node. Unlike hot threads this does nto sample and captures current state only.
```
Comment (ywangd) [inline server/src/main/java/org/elasticsearch/monitor/jvm/HotThreads.java:109]:
```suggestion
     * Capture and log the current threads on the local node. Unlike hot threads this does not sample and captures current state only.
```

File: server/src/main/java/org/elasticsearch/monitor/jvm/HotThreads.java [hunk 3]
```diff
@@ -104,6 +105,34 @@ public static void logLocalHotThreads(Logger logger, Level level, String prefix,
...
+     * @param referenceDocs A link to the docs describing how to decode the logging.
+     */
+    public static void logLocalCurrentThreads(Logger logger, Level level, String prefix, ReferenceDocs referenceDocs) {
+        if (logger.isEnabled(level) == false) {
+            return;
+        }
+
+        try (var writer = new StringWriter()) {
+            new HotThreads().busiestThreads(500).threadElementsSnapshotCount(1).ignoreIdleThreads(false).detect(writer, () -> {
+                logger.log(level, "{}: {}", prefix, writer.toString(), referenceDocs);
```
Comment (ywangd) [inline server/src/main/java/org/elasticsearch/monitor/jvm/HotThreads.java:125]: Are we ignoring referenceDocs here? The message has only two placeholders. I do think logging referenceDocs for every thread is excessive. But I am not sure the intention here.

File: server/src/main/java/org/elasticsearch/monitor/jvm/HotThreads.java [hunk 4]
```diff
@@ -104,6 +105,34 @@ public static void logLocalHotThreads(Logger logger, Level level, String prefix,
...
+     * @param prefix        The prefix to emit on each chunk of the logging.
+     * @param referenceDocs A link to the docs describing how to decode the logging.
+     */
+    public static void logLocalCurrentThreads(Logger logger, Level level, String prefix, ReferenceDocs referenceDocs) {
+        if (logger.isEnabled(level) == false) {
+            return;
+        }
+
+        try (var writer = new StringWriter()) {
+            new HotThreads().busiestThreads(500).threadElementsSnapshotCount(1).ignoreIdleThreads(false).detect(writer, () -> {
```
Comment (ywangd) [inline server/src/main/java/org/elasticsearch/monitor/jvm/HotThreads.java:124]: I wonder whether we should `ignoreIdleThreads(true)`? Including them provides a complete picture but I often found skipping them (like 90% of the output) is quite painful. Since this new method is meant for mostly internal investigation and per thread, I wonder whether we could do away with idle threads.

File: server/src/main/java/org/elasticsearch/monitor/jvm/HotThreads.java (hunk 1 above)
Comment (ywangd) [reply server/src/main/java/org/elasticsearch/monitor/jvm/HotThreads.java:285]: Good idea! Wrapping the sleep inside a `if (j < threadElementsSnapshotCount - 1)` check might be a smaller changeset.

File: server/src/main/java/org/elasticsearch/monitor/jvm/HotThreads.java (hunk 4 above)
Comment (idegtiarenko) [reply server/src/main/java/org/elasticsearch/monitor/jvm/HotThreads.java:124]: Good point. This was copied from 
https://github.com/elastic/elasticsearch/blob/7344b151fb323a4f106fe5a5819b49f70fc35278/server/src/main/java/org/elasticsearch/monitor/jvm/HotThreads.java#L102
but I also do not find them very informative.

@DaveCTurner was there a reason to enable them explicitly?

Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/monitor/jvm/HotThreads.java:124]: We can drop them here, sure.

File: server/src/main/java/org/elasticsearch/monitor/jvm/HotThreads.java (hunk 3 above)
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/monitor/jvm/HotThreads.java]: The point of the `ReferenceDocs` link was largely so that we had somewhere to write instructions about how to decode these stack traces. Not really necessary if we're not encoding them in the first place.

File: server/src/main/java/org/elasticsearch/monitor/jvm/HotThreads.java (hunk 1 above)
Comment (pxsalehi) [reply server/src/main/java/org/elasticsearch/monitor/jvm/HotThreads.java:285]: nit: maybe instead adding an if in the for loop is a bit more readable.

Comment (idegtiarenko) [reply server/src/main/java/org/elasticsearch/monitor/jvm/HotThreads.java:284]: I prefer to make a loop one iteration shorter rather than every time check if it is the last iteration.

File: server/src/main/java/org/elasticsearch/monitor/jvm/HotThreads.java [hunk 5]
```diff
@@ -191,11 +219,12 @@ public HotThreads sortOrder(SortOrder order) {
...
+        detect(writer, () -> {});
+    }
+
+    public void detect(Writer writer, Runnable onNextThread) throws Exception {
         synchronized (mutex) {
-            innerDetect(ManagementFactory.getThreadMXBean(), SunThreadInfo.INSTANCE, Thread.currentThread().getId(), (interval) -> {
-                Thread.sleep(interval);
-                return null;
-            }, writer);
+            innerDetect(ManagementFactory.getThreadMXBean(), SunThreadInfo.INSTANCE, Thread.currentThread().getId(), writer, onNextThread);
```
Comment (ywangd) [inline server/src/main/java/org/elasticsearch/monitor/jvm/HotThreads.java:227]: The threadSleep function is irrelevant to this change. Should we keep it?

File: server/src/main/java/org/elasticsearch/monitor/jvm/HotThreads.java [hunk 6]
```diff
@@ -104,6 +105,33 @@ public static void logLocalHotThreads(Logger logger, Level level, String prefix,
...
+    public static void logLocalCurrentThreads(Logger logger, Level level, String prefix) {
+        if (logger.isEnabled(level) == false) {
+            return;
+        }
+
+        try (var writer = new StringWriter()) {
+            new HotThreads().busiestThreads(500).threadElementsSnapshotCount(1).detect(writer, () -> {
+                logger.log(level, "{}: {}", prefix, writer.toString());
+                writer.getBuffer().setLength(0);
+            });
```
Comment (ywangd) [inline server/src/main/java/org/elasticsearch/monitor/jvm/HotThreads.java:126]: I wonder whether it is worthwhile to add similar counting and summary of total number of threads similar to how hot-threads are logged. We don't really need the counting to preserve orders for decompressing, but it seems useful to have a summary on how many threads are logged?

**Review by ywangd — COMMENTED** [review]
I think we could use a minimal test for it similar to `HotThreadsIT#testLogLocalHotThreads`?

File: server/src/main/java/org/elasticsearch/monitor/jvm/HotThreads.java [hunk 7]
```diff
@@ -104,6 +105,33 @@ public static void logLocalHotThreads(Logger logger, Level level, String prefix,
...
+    /**
+     * Capture and log the current threads on the local node. Unlike hot threads this does not sample and captures current state only.
+     * Useful for capturing stack traces for unexpectedly-slow operations in production. The resulting message might be large, so it is
+     * split per thread and logged as multiple entries.
+     *
+     * @param logger        The logger to use for the logging
+     * @param level         The log level to use for the logging.
+     * @param prefix        The prefix to emit on each chunk of the logging.
+     */
+    public static void logLocalCurrentThreads(Logger logger, Level level, String prefix) {
```
Comment (pxsalehi) [inline server/src/main/java/org/elasticsearch/monitor/jvm/HotThreads.java:117]: what was the specific use case that would benefit from this? Maybe we could use it there with this PR too? otherwise this seems unused.

File: server/src/main/java/org/elasticsearch/monitor/jvm/HotThreads.java [hunk 8]
```diff
@@ -296,10 +322,11 @@ void innerDetect(
             .append(", ignoreIdleThreads=")
             .append(Boolean.toString(ignoreIdleThreads))
             .append(":\n");
+        onNextThread.run();
 
         // Capture before and after thread state with timings
         Map<Long, ThreadTimeAccumulator> previousThreadInfos = getAllValidThreadInfos(threadBean, sunThreadInfo, currentThreadId);
-        threadSleep.apply(interval.millis());
+        Thread.sleep(interval.millis());
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/monitor/jvm/HotThreads.java:329]: This slows the tests down doesn't it? Previously we were not actually sleeping here at all, whereas now we do. Is this change needed for this utility?

File: server/src/main/java/org/elasticsearch/monitor/jvm/HotThreads.java (hunk 6 above)
Comment (ywangd) [reply server/src/main/java/org/elasticsearch/monitor/jvm/HotThreads.java:126]: We synced offline and agreed that search with prefix (likely unqiue) should be able to get us all the relevant entries without counting and summary. They might still be useful if somehow ingestion had problem to take in all data. But that should be a rare case and we can have follow-ups if we really need them.

**Review by ywangd — APPROVED** [review]
LGTM as long as CI is happy. Thanks!

File: server/src/main/java/org/elasticsearch/monitor/jvm/HotThreads.java (hunk 7 above)
Comment (idegtiarenko) [reply server/src/main/java/org/elasticsearch/monitor/jvm/HotThreads.java:118]: Correct, it is currently unused. The goal of the pr is to introduce the mechanism to log hot threads in plain threads without changing existing behavior. Existing `logLocalHotThreads` usages could be moved to `logLocalCurrentThreads` eventually. Meanwhile this change could help with serverless performance investigations.

File: server/src/main/java/org/elasticsearch/monitor/jvm/HotThreads.java [hunk 9]
```diff
@@ -104,6 +105,33 @@ public static void logLocalHotThreads(Logger logger, Level level, String prefix,
...
+     * @param prefix        The prefix to emit on each chunk of the logging.
+     */
+    public static void logLocalCurrentThreads(Logger logger, Level level, String prefix) {
+        if (logger.isEnabled(level) == false) {
+            return;
+        }
+
+        try (var writer = new StringWriter()) {
+            new HotThreads().busiestThreads(500).threadElementsSnapshotCount(1).detect(writer, () -> {
+                logger.log(level, "{}: {}", prefix, writer.toString());
```
[the remark this answers is not in the document]
Comment (idegtiarenko) [reply server/src/main/java/org/elasticsearch/monitor/jvm/HotThreads.java:125]: This is technically possible but would require much bigger refactoring to introduce domain model for the stack trace, thread name, cpu usage as well as various output modes (such as MEM). I would like to leave it outside of the scope of this change.
